From 84dd621ddffc053654119a76bc8caa98436b75a1 Mon Sep 17 00:00:00 2001
From: Michal Schmidt <mschmidt@redhat.com>
Date: Fri, 5 Sep 2014 19:49:30 -0400
Subject: [ethernet] bnx2x: Don't show port statistics for VFs

Message-id: <1409946615-5862-26-git-send-email-mschmidt@redhat.com>
Patchwork-id: 91568
O-Subject: [RHEL7.1 PATCH 25/70] bnx2x: Don't show port statistics for VFs
Bugzilla: 1089402
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1089402

Upstream commit:
commit d8361051000f27d2c9467c1f18985f9ce2123415
Author: Yuval Mintz <Yuval.Mintz@qlogic.com>
Date:   Sun Mar 23 18:12:26 2014 +0200

    bnx2x: Don't show port statistics for VFs

    VFs are currently showing port statistics, although they can't really access
    those - thus all such statistics will always show a value of 0.

    This patch removes said statistics from the VF's view as to not confuse the
    user.

    Signed-off-by: Yuval Mintz <Yuval.Mintz@qlogic.com>
    Signed-off-by: Ariel Elior <Ariel.Elior@qlogic.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_ethtool.c b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_ethtool.c
index 38fc794..b6de05e 100644
--- a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_ethtool.c
+++ b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_ethtool.c
@@ -2969,8 +2969,9 @@ static void bnx2x_self_test(struct net_device *dev,
 #define IS_PORT_STAT(i) \
  ((bnx2x_stats_arr[i].flags & STATS_FLAGS_BOTH) == STATS_FLAGS_PORT)
 #define IS_FUNC_STAT(i)  (bnx2x_stats_arr[i].flags & STATS_FLAGS_FUNC)
-#define IS_MF_MODE_STAT(bp) \
-   (IS_MF(bp) && !(bp->msg_enable & BNX2X_MSG_STATS))
+#define HIDE_PORT_STAT(bp) \
+  ((IS_MF(bp) && !(bp->msg_enable & BNX2X_MSG_STATS)) || \
+   IS_VF(bp))
 
 /* ethtool statistics are displayed for all regular ethernet queues and the
  * fcoe L2 queue if not disabled
@@ -2992,7 +2993,7 @@ static int bnx2x_get_sset_count(struct net_device *dev, int stringset)
           BNX2X_NUM_Q_STATS;
   } else
    num_strings = 0;
-  if (IS_MF_MODE_STAT(bp)) {
+  if (HIDE_PORT_STAT(bp)) {
    for (i = 0; i < BNX2X_NUM_STATS; i++)
     if (IS_FUNC_STAT(i))
      num_strings++;
@@ -3047,7 +3048,7 @@ static void bnx2x_get_strings(struct net_device *dev, u32 stringset, u8 *buf)
   }
 
   for (i = 0, j = 0; i < BNX2X_NUM_STATS; i++) {
-   if (IS_MF_MODE_STAT(bp) && IS_PORT_STAT(i))
+   if (HIDE_PORT_STAT(bp) && IS_PORT_STAT(i))
     continue;
    strcpy(buf + (k + j)*ETH_GSTRING_LEN,
        bnx2x_stats_arr[i].string);
@@ -3105,7 +3106,7 @@ static void bnx2x_get_ethtool_stats(struct net_device *dev,
 
  hw_stats = (u32 *)&bp->eth_stats;
  for (i = 0, j = 0; i < BNX2X_NUM_STATS; i++) {
-  if (IS_MF_MODE_STAT(bp) && IS_PORT_STAT(i))
+  if (HIDE_PORT_STAT(bp) && IS_PORT_STAT(i))
    continue;
   if (bnx2x_stats_arr[i].size == 0) {
    /* skip this counter */
-- 
1.7.1